File: ATV8_TIPOS_ABSTRATOS/haskell/livro.hs
newtype Livro = Livro (String, String, Int, Int)

inicializar :: String -> String -> Int -> Int -> Livro
inicializar titulo autor ano copias = Livro (titulo autor ano copias)

emprestaCopia :: Livro -> Livro
emprestaCopia (Livro titulo autor ano copias) = Livro (titulo autor ano (copias - 1))

devolveCopia :: Livro -> Livro
devolveCopia (Livro (titulo autor ano copias)) = (Livro titulo autor ano (copias + 1))

verificaCopia :: Livro -> Int
devolveCopia (Livro (_, _, _, copias)) = copias

getLivro :: Livro -> String
getLivro (Livro (titulo, autor, ano, _)) = "Título: " ++ titulo ++ ", Autor: " ++ autor ++ ", Ano: " ++ show ano

File: ATV8_TIPOS_ABSTRATOS/java/Livro.java
package ATV8_TIPOS_ABSTRATOS.java;

public class Livro {
    private String titulo;
    private String autor;
    private int ano;
    private int copias;

    public Livro(String titulo, String autor, int ano, int copias) {
        this.titulo = titulo;
        this.autor = autor;
        this.ano = ano;
        this.copias = copias;
    }

    public void emprestaCopia() {
        this.copias--;
    }

    public void devolveCopia() {
        this.copias++;
    }

    public int verificaCopia() {
        return this.copias;
    }

    @Override
    public String toString() {
        return "Título: " + this.titulo + "\nAutor: " + this.autor + "\nAno: " + this.ano;
    }
}